picard/util/pipe.py

```diff
@@ -0,0 +1,232 @@
+
+                        self.__create_unix_pipe()
+                        break
+            else:
+                self.__create_unix_pipe()
+
+    def __pipe_exists(self) -> bool:
+        return os.path.exists(self.path)
+
+    def __generate_filename(self) -> str:
+        prefix: str = ""
+        username: str = os.getlogin()
+        if self.__is_win:
+            prefix = "\\\\.\\pipe\\"
+        else:
+            prefix = "/tmp/"
+
+        return f"{prefix}{self.__app_name}_v{self.__app_version}_{username}_pipe_file"
+
+    def __create_unix_pipe(self) -> None:
+        os.mkfifo(self.path)
+        self.is_pipe_owner = True
+
+    def __win_sender(self, message: str) -> bool:
+        pipe = win32pipe.CreateNamedPipe(
+            self.path,
+            win32pipe.PIPE_ACCESS_DUPLEX,
+            win32pipe.PIPE_TYPE_MESSAGE | win32pipe.PIPE_WAIT,
+            self.__MAX_INSTANCES,
+            self.__OUT_BUFFER_SIZE,
+            self.__IN_BUFFER_SIZE,
+            self.__DEFAULT_TIMEOUT,
+            None)
+        try:
+            win32pipe.ConnectNamedPipe(pipe, None)
+            win32file.WriteFile(pipe, str.encode(message))
+        finally:
+            win32file.CloseHandle(pipe)
+
+        return True
+
+    def __unix_sender(self, message: str) -> bool:
+        with open(self.path, 'a') as fifo:
+            fifo.write(message)
+        return True
+
+    def send_to_pipe(self, message: str, timeout_secs: float = 1.5) -> bool:
+        __pool = concurrent.futures.ThreadPoolExecutor()
+        sender = None
+        if self.__is_win:
+            sender = __pool.submit(self.__win_sender, message)
+        else:
+            sender = __pool.submit(self.__unix_sender, message)
+
+        try:
+            if sender.result(timeout=timeout_secs):
+                return True
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the sender
+            self.read_from_pipe()
+
+        return False
+
+    def read_from_pipe(self, timeout_secs: float = 1.5) -> str:
+        if self.__is_win:
+            return str(self.__read_from_win_pipe(timeout_secs))
+        else:
+            return self.__read_from_unix_pipe(timeout_secs)
+
+    def __win_reader(self) -> str:
+        response = ""  # type: ignore
+
+        try:
+            pipe = win32file.CreateFile(
+                self.path,
+                win32file.GENERIC_READ | win32file.GENERIC_WRITE,
+                self.__SHARE_MODE,
+                None,
+                win32file.OPEN_EXISTING,
+                self.__FLAGS_AND_ATTRIBUTES,
+                None
+            )
+            while not response:
+                response = win32file.ReadFile(pipe, 64 * 1024)
+
+        except WinApiError as err:
+            if err.args[0] == 2:
+                raise FileNotFoundError(Pipe.NOT_FOUND_MESSAGE)
+            elif err.args[0] == 109:
```

Comment:
Perhaps use constants as well for those error codes.

Reply:
Ok, they'll be in the same place with a reference article as well.